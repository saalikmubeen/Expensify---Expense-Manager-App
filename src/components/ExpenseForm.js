import React, { Component } from 'react'
import moment from "moment";
import 'react-dates/initialize'
import {SingleDatePicker} from "react-dates";
import 'react-dates/lib/css/_datepicker.css';

class ExpenseForm extends Component {
    constructor(props){
        super(props);
        this.state = {
            description: this.props.expense ? this.props.expense.description : "",
            note: this.props.expense ? this.props.expense.note : "",
            amount: this.props.expense ? this.props.expense.amount : "",
            createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
            calenderFocused: false,
            error: ""
        }
    }

    handleChange = (e) => {
        this.setState({[e.target.name]: e.target.value});
    }

    handleSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            return this.setState({error: "Provide a description and amount!"})
        }

        this.setState({error: ""});
        this.props.saveExpense({
             description: this.state.description, 
             note: this.state.note, 
             amount: this.state.amount, 
             createdAt: this.state.createdAt.valueOf()})
    }

    render() {
        return (
            
            <form className="form" onSubmit={this.handleSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="Description"
                    name="description"        
                    autoFocus
                    autoComplete="off"
                    className="text-input"
                    value={this.state.description}
                    onChange={this.handleChange}
                />
                
                <input
                    type="text"
                    placeholder="Amount"
                    name="amount"
                    autoComplete="off"
                    className="text-input"
                    value={this.state.amount}
                    onChange={this.handleChange}
                />

                <SingleDatePicker
                    date={this.state.createdAt} // momentPropTypes.momentObj or null
                    onDateChange={date => this.setState({ createdAt: date })} // PropTypes.func.isRequired
                    focused={this.state.calenderFocused} // PropTypes.bool
                    onFocusChange={({ focused }) => this.setState({ calenderFocused: focused })} // PropTypes.func.isRequired
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />

                <textarea
                    placeholder="Add a note for your expense (optional)"
                    name="note"
                    className="textarea"
                    value={this.state.note}
                    onChange={this.handleChange}
                >
                </textarea>

                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
            
        )
    }
}


export default ExpenseForm;