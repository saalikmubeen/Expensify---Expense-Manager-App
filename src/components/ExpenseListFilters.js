import React, { Component } from 'react'
import {connect} from "react-redux";
import 'react-dates/initialize'
import {DateRangePicker} from "react-dates";
import 'react-dates/lib/css/_datepicker.css';

import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from "../actions/filters";


class ExpenseListFilters extends Component {
       constructor(props){
           super(props);

           this.state = {
               focusedInput: false
           }
       }

    handleChange = (e) => {
         this.props.dispatch(setTextFilter(e.target.value))
    }

    handleFilterByChange = (e) => {
         if(e.target.value === "date"){
             this.props.dispatch(sortByDate())
         }else if(e.target.value === "amount"){
             this.props.dispatch(sortByAmount())
         }
    }

    handleDatesChange = ({startDate, endDate}) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    }

    render() {
        return (

            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            type="text"
                            className="text-input"
                            placeholder="Search expenses"
                            value={this.props.filters.text}
                            onChange={this.handleChange}
                        />
                    </div>
        
                    <div className="input-group__item">
                        <select
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.handleFilterByChange}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
        
                    <div className="input-group__item">
                        <DateRangePicker
                                startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                                startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                                endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                                endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                                onDatesChange={this.handleDatesChange} // PropTypes.func.isRequired,
                                focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                                onFocusChange={focusedInput => this.setState({ focusedInput })} // PropTypes.func.isRequired,
                                numberOfMonths={1}
                                isOutsideRange={() => false}
                                showClearDates={true}
                            />   
                    </div>
                </div>
            </div>
        )
    }
}

var mapStateToProps = (state) => {
    return {filters: state.filters}
}

export default connect(mapStateToProps)(ExpenseListFilters);







// class ExpenseListFilters extends Component {

//     state = {
//         text: this.props.filters.text
//     }

//     handleChange = (e) => {
//         this.setState({text: e.target.value}, () => {
//             this.props.dispatch(setTextFilter(this.state.text))
//         })
//     }

//     render() {
//         console.log(this.props.filters.text)
//         return (
//             <div>
//                 <input type="text" value={this.state.text} onChange={this.handleChange}/>
//             </div>
//         )
//     }
// }


