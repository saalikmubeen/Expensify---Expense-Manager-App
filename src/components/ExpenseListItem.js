import React from 'react';
import {Link} from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

class ExpenseListItem extends React.Component{

        render(){
            var {description, amount, createdAt, id} = this.props
            return (
                
              <Link className="list-item" to={`/edit/${id}`}>
                  <div>
                    <h3 className="list-item__title">{description}</h3>
                    <span className="list-item__sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</span>
                  </div>
                  <h3 className="list-item__data">{numeral(amount).format('$0,0.00')}</h3>
              </Link>
                
        )
    }
}


export default ExpenseListItem;