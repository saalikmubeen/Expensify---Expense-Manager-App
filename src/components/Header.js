import React, { Component } from 'react'
import {NavLink} from "react-router-dom";
import {connect} from 'react-redux';
import {logoutUser} from "../actions/auth";


class Header extends Component {
    render() {
        return (

            <header className="header">
                <div className="content-container">
                  <div className="header__content">
                    <NavLink className="header__title" to="/dashboard">
                      <h1>Expensify</h1>
                    </NavLink>
                    <button className="button button--link" onClick={this.props.logoutUser}>Logout</button>
                  </div>
                </div>
            </header>
        )
    }
}

var matchDispatchToProps = function(dispatch){
    return ({
        logoutUser: () => dispatch(logoutUser())
    })
}

export default connect(undefined, matchDispatchToProps)(Header);