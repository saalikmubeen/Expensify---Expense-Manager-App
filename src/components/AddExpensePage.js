import React, { Component } from 'react'
import {connect} from 'react-redux';
import {asyncAddExpense} from "../actions/expenses";
import ExpenseForm from "./ExpenseForm"

class AddExpensePage extends Component {

    saveExpense = async (expenseObj) => {
        await this.props.dispatch(asyncAddExpense(expenseObj))
        this.props.history.push("/");
    }

    render() {
        return (

            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>

                <div className="content-container">
                    <ExpenseForm saveExpense={this.saveExpense}/>
                </div>
            </div>
        )
    }
}

export default connect()(AddExpensePage);