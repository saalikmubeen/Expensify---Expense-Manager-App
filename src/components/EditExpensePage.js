import React, { Component } from 'react'
import {connect} from "react-redux";
import {asyncEditExpense, asyncRemoveExpense} from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

class EditExpensePage extends Component {

    editExpense = (expenseObj) => {
        this.props.dispatch(asyncEditExpense(this.props.expense.id, expenseObj))
                .then(() => {
                    this.props.history.push("/");
                })
    }

    handleClick = () => {
         this.props.dispatch(asyncRemoveExpense({id: this.props.expense.id}))
                  .then(() => {
                   this.props.history.push("/");
                   })
    }

    render() {
        return (

            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>

                <div className="content-container">
                    <ExpenseForm expense={this.props.expense} saveExpense={this.editExpense}/>
                    <button className="button button--secondary" onClick={this.handleClick}>Remove Expense</button>
                </div>
            </div>
        )
    }
}

var matchStateToProps = (state, props) => {
         return {
             expense: state.expenses.find((expense) => expense.id === props.match.params.id)
         }
}

export default connect(matchStateToProps)(EditExpensePage);