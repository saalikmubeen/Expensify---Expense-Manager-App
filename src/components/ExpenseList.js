import React from 'react'
import {connect} from 'react-redux';
import ExpenseListItem from "./ExpenseListItem";
import getVisibleExpenses from "../selectors/expenses";

var ExpenseList = (props) => {
    return(

        <div className="content-container">
            <div className="list-header">
                <div className="show-for-mobile">Expenses</div>
                <div className="show-for-desktop">Expense</div>
                <div className="show-for-desktop">Amount</div>
            </div>

            <div className="list-body">
            {
            props.expenses.length === 0 ? (
                <div className="list-item list-item--message">
                <span>No expenses</span>
                </div>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense} />;
                })
                )
            }
            </div>
        </div>
    )
}

var mapStateToProps = (state) => {
    return {expenses: getVisibleExpenses(state.expenses, state.filters)}
}


export default connect(mapStateToProps)(ExpenseList);

// var ConnectedExpenseList = connect((state) => {
//     return {
//        ...state
//     }
// })(ExpenseList)

// export default ConnectedExpenseList;

// connect takes a mapStateToProps(a function) as argument that passes the store state/data
// to the component that needs to connect to the store. The store data passed to the component 
// Every time store state changes connect runs and passes fresh store state to its connected components
// This makes our components to be in sync with the store.
// if we don't pass any arguments to connect() then only dispatch will be accessible in 
// our component via props.dispatch we won't be able to access store state/data. 

