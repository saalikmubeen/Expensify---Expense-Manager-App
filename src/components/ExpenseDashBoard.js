import React from 'react';
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpensesSummary from "./ExpensesSummary";

var ExpenseDashBoard = (props) => {
    return (
        <div>
            <ExpensesSummary/>
            <ExpenseListFilters/>
            <ExpenseList/>
        </div>
    )
}

export default ExpenseDashBoard;