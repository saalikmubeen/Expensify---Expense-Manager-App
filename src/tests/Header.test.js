import React from "react";
import { shallow } from "enzyme";
import { Header } from "../components/Header";

test("Expect to render Header component", () => {
    expect.assertions(2);
    const wrapper = shallow(<Header />);
    expect(wrapper.find("h1").text()).toBe("Expensify");
    expect(wrapper).toMatchSnapshot();
});
