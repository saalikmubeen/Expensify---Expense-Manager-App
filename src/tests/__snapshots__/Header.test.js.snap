// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect to render Header component 1`] = `ShallowWrapper {}`;
